### DISCUSSION TITLE
Using libFLAC on ESP32 with PCM5102A for 44.1/16/2 audio

### DISCUSSION THREAD

**Discussion by geoffw64** [pschatzmann/arduino-audio-tools#1596] [Q&A] [answered]
Hi Phil, Hope you're well. 
I'm using the libFLAC library you very kindly ported for Arduino IDE :) 

My setup is Adafruit Huzzah32 with ESP32 WROOM32 and an Adafruit Adalogger with SanDisk Ultra Extreme Pro SD card.  I have connected a PCM5102A DAC board and have soldered the FMT jumper to LOW.  
A sketch sending sine wave samples to the DAC from *memory* at 44.1kHz 16-bit 2 channel produces clean sound, so I think my hardware setup is ok.  
However, when I read a 44.1kHz 16-bit 2 channel flac audio file from an SD card and decode it using libFLAC,  the sound is choppy and has artefacts. 
I tested the speed of the SD card in a stand-alone sketch and it reads the entire 26MB audio file in 71 seconds, reading chunks into an 8KB buffer.  This is roughly 370 Kbytes/sec or 3Mbps; much faster than the actual playback time.  
Questions..
1. Have you found that you need to pre-read audio data from files on SD into memory so that the decoder isn't waiting each time is runs read_callback?  
2. Does anything look astray with my i2s config or libFLAC callbacks below?   
  i2s_config_t i2s_config = {
    .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_TX),
    .sample_rate = SAMPLE_RATE, 
    .bits_per_sample = (i2s_bits_per_sample_t)BITS_PER_SAMPLE, 
    .channel_format = (i2s_channel_fmt_t)I2S_CHANNEL_FMT_RIGHT_LEFT,  // 2 channels
    .communication_format = (i2s_comm_format_t)I2S_COMM_FORMAT_I2S,   // Set for external DAC
    .intr_alloc_flags = ESP_INTR_FLAG_LEVEL2,                         // Set interrupt priority
    .dma_buf_count = BUFFER_COUNT,
    .dma_buf_len = BUFFER_LEN, 
    .use_apll = false   // Does not use audio clock
  }; 

Here are my read and write callbacks with relevant definitions:  
#define SAMPLE_RATE         44100             // Samples per second (per channel)
#define BITS_PER_SAMPLE I2S_BITS_PER_SAMPLE_16BIT // 16 bits per sample
#define BUFFER_COUNT        2                 // Number of DMA buffers 
#define NUM_CHANNELS        2                 // Stereo
#define SAMPLE_LEN_BYTES    (BITS_PER_SAMPLE / 8) // Size in bytes of a single sample for one channel
#define SAMPLES_PER_BUFFER  4                // Number of samples to be written to buffer for each channel 
#define BUFFER_ELEMENTS     (SAMPLES_PER_BUFFER * NUM_CHANNELS) // Number of elements needed to hold contents of buffer
#define BUFFER_LEN          (BUFFER_ELEMENTS * SAMPLE_LEN_BYTES)  // Length of each DMA buffer in bytes 
int16_t i2s_buffer_16[BUFFER_ELEMENTS];     // Buffer used to hold 16-bit signed sample values 

FLAC__StreamDecoderReadStatus read_callback(
        const FLAC__StreamDecoder* decoder,
        FLAC__byte flac_buffer[], 
        size_t* bytes, 
        void* client_data) { 
  if (final) {
    *bytes = 0;       // Zero bytes passed to decoder
    final = false; 
    return FLAC__STREAM_DECODER_READ_STATUS_END_OF_STREAM; 
  } 
  // Executes if not final
  *bytes = file.read(flac_buffer, *bytes); // Read bytes up to provided buffer size from file
  if (!file.available()) {
    final = true;           // End of file reached
  }
  return FLAC__STREAM_DECODER_READ_STATUS_CONTINUE; 
} 
FLAC__StreamDecoderWriteStatus write_callback(
        const FLAC__StreamDecoder* decoder, 
        const FLAC__Frame* frame, 
        const FLAC__int32* const flac_buffer[], 
        void* client_data) { 
  FLAC__FrameHeader header = frame -> header; 
  uint32_t blocksize = header.blocksize; 
  const FLAC__int32* audio_samples_ptr_0 = flac_buffer[0]; // Pointer holding address of decoded samples for channel 0
  const FLAC__int32* audio_samples_ptr_1 = flac_buffer[1]; // Pointer holding address of decoded samples for channel 1
  const FLAC__int32* end_of_buffer_ptr = flac_buffer[0] + blocksize; // Pointer to end of channel 0 sample array
  while (audio_samples_ptr_0 != end_of_buffer_ptr) { // Check for end of array; assumes same number of samples in each channel
    FLAC__int32 sample = *audio_samples_ptr_1 / 8; // Attenuate the original sample. 
    i2s_buffer_16[buf_el] = i2s_buffer_16[buf_el + 1] = (int16_t)sample; // Set value for channel 0 and 1
    if (buf_el == BUFFER_ELEMENTS) { 
      write_i2s_buffer();               // Buffer is full; write it out
      buf_el = 0;                       // Reset counter to start filling buffer again
    } else 
      buf_el += 2;                      // Recognise 2 samples added to buffer 

    ++audio_samples_ptr_0;              // Increment both pointers along the arrays
    ++audio_samples_ptr_1;
  } // end while
  return FLAC__STREAM_DECODER_WRITE_STATUS_CONTINUE; 
}

I run the decoding and i2s processing in a high priority task.  APPLICATION_CPU is defined as 1
xTaskCreatePinnedToCore(audio_task, "audio_task", 8 * 1024, NULL, configMAX_PRIORITIES - 1, NULL, APPLICATION_CPU);

Sorry for the long post! and thanks in advance!  
Cheers, Geoff.

**Comment by pschatzmann** [comment, marked as answer]
I suggest to compare your result with [my example sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-sd_flac-audiokit/streams-sd_flac-audiokit.ino): Just replace the AudioBoardStream with an regular [I2SStream](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC).

If this is giving the same bad result you will need to check the encoing of your files: I think a level of 0 (fastest) or 1 gives the best result and an ESP32 might not be powerfull enough to handle 8 (best quality).

Using the AudioTools has the advantage that you don't need to rewrite your sketch if you move to the latest ESP 3.0.1 core version.

**Reply by geoffw64** [reply]
Thanks Phil, I appreciate your time. 
The example sketch does produce better sound so something is not right with my sketch.  I can only listen to a couple of seconds though, as my power amp has no volume control  :) .  To listen for longer I'll need to reduce volume.  I set about using VolumeStream and read through an example.  
The FLACDecoder isn't accepted by the EncodedAudioStream  constructor.  I guess this is because it's a decoder not an audio stream.  Do I need to decode the FLAC to I2S using StreamCopy first, and then  feed that stream into EncodedAudioStream?  I looked through the examples but couldn't find one with a similar use case.  
Your audio tools are awesome.  I can see you've done a lot of work.  The only reason I'm writing my own FLAC player is that I plan to port it to Raspberry Pi as part of a broader project.

**Comment by pschatzmann** [comment]
To add volume control, just wrap I2S with a VolumeStream and use the VolumeStream as output.

You can compile an AudioTools sketch on Rasperrry Pi as well...

**Reply by geoffw64** [reply]
Ah, so simple, thanks.  That works very nicely and the sound is actually pretty good through my ICEpower amplifier and KEF Q10 speakers in my office.  No dropouts at all playing through a track.  
Thanks for letting me know that AudioTools compiles on Pi.  I will give it a try when I've done a bit more on the ESP32.

**Comment by geoffw64** [comment]
Coffee is on the way  :)

**Comment by geoffw64** [comment]
Hey Phil, 
Just wondering why you haven't implemented the FLAC__stream_decoder_finish() function in CodecFLAC? 
This seems to be the recommended way to decode multiple files eg if playing a series of songs from a list.  The function de-initializes the decoder, so you can call initialize again. 
Or is there another way to decode multiple flac files without constructing a new decoder each time?  
Cheers, 
Geoff.

**Reply by pschatzmann** [reply]
It is my expectation that you can just assign a new file and continue with the copy()

**Reply by geoffw64** [reply]
Thought I tried that but perhaps I called begin() again.  I shall  test.  
Cheers, Geoff.

**Reply by geoffw64** [reply]
Hi Phil, 
Tried assigning a new file and continuing to copy.  Here is the output: 
VolumeStream.h : 199 - setVolume: 1.000000 at 0
09:33:22.500 -> [I] VolumeStream.h : 199 - setVolume: 1.000000 at 1
09:33:22.500 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:33:22.500 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:33:22.500 -> [I] CodecFLAC.h : 85 - FLAC is open
09:33:22.500 -> [I] AudioTypes.h : 127 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:33:22.500 -> [I] AudioTypes.h : 127 -  sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:33:22.500 -> [I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
09:33:22.500 -> [I] I2SConfigESP32.h : 81 - port_no: 0
09:33:22.500 -> [I] I2SConfigESP32.h : 82 - is_master: Master
09:33:22.500 -> [I] I2SConfigESP32.h : 83 - sample rate: 44100
09:33:22.500 -> [I] I2SConfigESP32.h : 84 - bits per sample: 16
09:33:22.500 -> [I] I2SConfigESP32.h : 85 - number of channels: 2
09:33:22.500 -> [I] I2SConfigESP32.h : 86 - signal_type: Digital
09:33:22.500 -> [I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
09:33:22.500 -> [I] I2SConfigESP32.h : 90 - auto_clear: true
09:33:22.500 -> [I] I2SConfigESP32.h : 92 - use_apll: true
09:33:22.500 -> [I] I2SConfigESP32.h : 97 - buffer_count:6
09:33:22.500 -> [I] I2SConfigESP32.h : 98 - buffer_size:512
09:33:22.500 -> [I] I2SConfigESP32.h : 103 - pin_bck: 14
09:33:22.500 -> [I] I2SConfigESP32.h : 105 - pin_ws: 15
09:33:22.500 -> [I] I2SConfigESP32.h : 107 - pin_data: 22
09:33:22.533 -> Now playing: /Music/03 Toto - Georgy Porgy.flac
09:33:22.533 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 0
09:33:22.533 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 1
09:37:41.009 -> [E] CodecFLAC.h : 109 - not active
09:37:41.042 -> Now playing: /Music/09 Toto - Hold The Line.flac
09:37:41.042 -> [E] CodecFLAC.h : 109 - not active
09:37:41.042 -> Now playing: /Music/01 Talking Heads - Psycho Killer.flac
09:37:41.042 -> [E] CodecFLAC.h : 109 - not active
09:37:41.042 -> Now playing: /Music/06 Steely Dan - Hey Nineteen.flac
09:37:41.042 -> [E] CodecFLAC.h : 109 - not active
09:37:41.042 -> Now playing: /Music/01 Dire Straits - So Far Away.flac
09:37:41.042 -> [E] CodecFLAC.h : 109 - not active

The decoder becomes inactive when it completes streaming the first flac file.  How can I continue playback without constructing a new decoder?  Should the ..._finish() function be implemented?  If you make a change I will test it.  
Here is my sketch: 

#include <SPI.h>
#include <SD.h>
#include "AudioTools.h"
#include "AudioI2S/I2SStream.h"
#include "AudioCodecs/CodecFLAC.h"

const int chipSelect=33;
I2SStream I2S_stream; 
I2SConfig I2S_config; 
VolumeStream volume_stream(I2S_stream); 
FLACDecoder dec;
File music_dir; 
File music_file;

void setup() {
  Serial.begin(115200);
  while (!Serial) 
    delay(1); 
  delay(200); 
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  
  // setup FLAC decoder
  //dec.setOutput(I2S_stream);  // Sends stream with volume unchanged
  volume_stream.begin(I2S_config); // Provides bits_per_sample and channels
  dec.setOutput(volume_stream); // Sends stream with reduced volume 
  dec.begin(); 
  // setup I2S before SD 
  I2S_config = I2S_stream.defaultConfig(TX_MODE);
  I2S_stream.begin(I2S_config); 
  // setup SD 
  SD.begin(chipSelect); 
  music_dir = SD.open("/Music"); 
  if (!music_dir) { 
    Serial.println("Could not open Music directory"); 
    while(1); 
  } 
}

void play(File music_file) {
  // Plays a music file
  dec.setInput(music_file);         // Set the input music file
  volume_stream.setVolume(0.12);    // Set volume
  while (dec.copy());               // Decode and send to DAC
}
void loop() {
  File music_file = music_dir.openNextFile();
  if (music_file) { 
    Serial.print("Now playing: "); 
    Serial.println(music_file.name()); 
    play(music_file); 
  } else {
    while(1);   // No more music to play
  }
}

**Comment by pschatzmann** [comment]
I [extended the begin() logic](https://github.com/pschatzmann/arduino-audio-tools/commit/9ce1898bcb6a82bc2837efc8cae625602f62dd88) a bit, so that the decoder gets closed with FLAC__stream_decoder_finish when it is open before calling FLAC__stream_decoder_init_stream.

So just call dec.begin() again after assigning a new file.

**Reply by geoffw64** [reply]
Hi Phil,  
Thanks for the update.  Unfortunately it doesn't work because is_active is (probably quite rightly) set to false in the read_callback when eof is reached:
    if (self->isEof(*bytes)){
        result = FLAC__STREAM_DECODER_READ_STATUS_END_OF_STREAM;
        self->is_active = false;
    }

Here is the Serial output:
09:27:41.594 -> [I] VolumeStream.h : 199 - setVolume: 1.000000 at 1
09:27:41.594 -> [I] AudioTypes.h : 127 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:27:41.594 -> [I] AudioTypes.h : 127 -  sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:27:41.594 -> [I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
09:27:41.594 -> [I] I2SConfigESP32.h : 81 - port_no: 0
09:27:41.594 -> [I] I2SConfigESP32.h : 82 - is_master: Master
09:27:41.594 -> [I] I2SConfigESP32.h : 83 - sample rate: 44100
09:27:41.594 -> [I] I2SConfigESP32.h : 84 - bits per sample: 16
09:27:41.594 -> [I] I2SConfigESP32.h : 85 - number of channels: 2
09:27:41.594 -> [I] I2SConfigESP32.h : 86 - signal_type: Digital
09:27:41.594 -> [I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
09:27:41.594 -> [I] I2SConfigESP32.h : 90 - auto_clear: true
09:27:41.594 -> [I] I2SConfigESP32.h : 92 - use_apll: true
09:27:41.594 -> [I] I2SConfigESP32.h : 97 - buffer_count:6
09:27:41.594 -> [I] I2SConfigESP32.h : 98 - buffer_size:512
09:27:41.594 -> [I] I2SConfigESP32.h : 103 - pin_bck: 14
09:27:41.594 -> [I] I2SConfigESP32.h : 105 - pin_ws: 15
09:27:41.594 -> [I] I2SConfigESP32.h : 107 - pin_data: 22
09:27:41.628 -> Now playing: /Music/03 Toto - Georgy Porgy.flac
09:27:41.628 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:27:41.628 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:27:41.628 -> [I] CodecFLAC.h : 85 - FLAC is open
09:27:41.628 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 0
09:27:41.628 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 1
09:32:00.055 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/09 Toto - Hold The Line.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/01 Talking Heads - Psycho Killer.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/06 Steely Dan - Hey Nineteen.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/01 Dire Straits - So Far Away.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:27:41.594 -> [I] VolumeStream.h : 199 - setVolume: 1.000000 at 1
09:27:41.594 -> [I] AudioTypes.h : 127 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:27:41.594 -> [I] AudioTypes.h : 127 -  sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:27:41.594 -> [I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
09:27:41.594 -> [I] I2SConfigESP32.h : 81 - port_no: 0
09:27:41.594 -> [I] I2SConfigESP32.h : 82 - is_master: Master
09:27:41.594 -> [I] I2SConfigESP32.h : 83 - sample rate: 44100
09:27:41.594 -> [I] I2SConfigESP32.h : 84 - bits per sample: 16
09:27:41.594 -> [I] I2SConfigESP32.h : 85 - number of channels: 2
09:27:41.594 -> [I] I2SConfigESP32.h : 86 - signal_type: Digital
09:27:41.594 -> [I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
09:27:41.594 -> [I] I2SConfigESP32.h : 90 - auto_clear: true
09:27:41.594 -> [I] I2SConfigESP32.h : 92 - use_apll: true
09:27:41.594 -> [I] I2SConfigESP32.h : 97 - buffer_count:6
09:27:41.594 -> [I] I2SConfigESP32.h : 98 - buffer_size:512
09:27:41.594 -> [I] I2SConfigESP32.h : 103 - pin_bck: 14
09:27:41.594 -> [I] I2SConfigESP32.h : 105 - pin_ws: 15
09:27:41.594 -> [I] I2SConfigESP32.h : 107 - pin_data: 22
09:27:41.628 -> Now playing: /Music/03 Toto - Georgy Porgy.flac
09:27:41.628 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:27:41.628 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:27:41.628 -> [I] CodecFLAC.h : 85 - FLAC is open
09:27:41.628 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 0
09:27:41.628 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 1
09:32:00.055 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/09 Toto - Hold The Line.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/01 Talking Heads - Psycho Killer.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/06 Steely Dan - Hey Nineteen.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active
09:32:00.165 -> Now playing: /Music/01 Dire Straits - So Far Away.flac
09:32:00.165 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:32:00.165 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:32:00.165 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:32:00.165 -> [E] CodecFLAC.h : 109 - not active

Relevant functions in my sketch: 
void play(File music_file) {
  // Plays a music file
  dec.setInput(music_file);         // Set the input music file
  dec.begin();                      // Start the decoder
  volume_stream.setVolume(0.12);    // Set volume
  while (dec.copy());               // Decode and send to DAC
}
void loop() {
  File music_file = music_dir.openNextFile();
  if (music_file) { 
    Serial.print("Now playing: "); 
    Serial.println(music_file.name()); 
    play(music_file); 
  } else {
    while(1);   // No more music to play
  }
}

**Comment by pschatzmann** [comment]
Ok, I see: I replaced the checking logic to use FLAC__stream_decoder_get_state() instead of is_active

**Reply by geoffw64** [reply]
Unfortunately this doesn't work.  
The read_callback correctly sets result to ...END_OF_STREAM but this doesn't set the decoder state to ...UNINITIALIZED.   
if (self->isEof(*bytes)){
        result = FLAC__STREAM_DECODER_READ_STATUS_END_OF_STREAM;
        self->is_active = false;
    } 

Looking at libFLAC, it seems to be necessary to call  ...DECODER_FINISH() to set the state to ...UNINITIALIZED .

**Comment by pschatzmann** [comment]
I don't quite understand ? 
I was running your test and for me it was working as expected:
```
11:04:53.148 -> [I] I2SConfigESP32V1.h : 89 - pin_data: 26
11:04:53.148 -> [I] I2SESP32V1.h : 201 - tx: 26, rx: -1
11:04:53.148 -> [I] VolumeStream.h : 199 - setVolume: 1.000000 at 0
11:04:53.180 -> [I] VolumeStream.h : 199 - setVolume: 1.000000 at 1
11:04:53.180 -> Now playing: test0.flac
11:04:53.180 -> [I] CodecFLAC.h : 62 - virtual bool audio_tools::FLACDecoder::begin()
11:04:53.212 -> [I] CodecFLAC.h : 94 - FLAC is open
11:04:53.212 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 0
11:04:53.212 -> [I] VolumeStream.h : 199 - setVolume: 0.120000 at 1
11:10:52.949 -> [E] CodecFLAC.h : 122 - not active
11:10:53.058 -> Now playing: test1.flac
11:10:53.058 -> [I] CodecFLAC.h : 62 - virtual bool audio_tools::FLACDecoder::begin()
11:10:53.058 -> [I] CodecFLAC.h : 94 - FLAC is open
11:16:52.802 -> [E] CodecFLAC.h : 122 - not active
11:16:52.802 -> Now playing: test2.flac
11:16:52.802 -> [I] CodecFLAC.h : 62 - virtual bool audio_tools::FLACDecoder::begin()
11:16:52.802 -> [I] CodecFLAC.h : 94 - FLAC is open
```

You can avoid the 'not active' Error message above with

```C++
while (dec) dec.copy();
```

**Reply by geoffw64** [reply]
Hmmm, I must be doing something stupid.  
With this code: 
void play(File music_file) {
  // Plays a music file
  dec.setInput(music_file);         // Set the input music file
  dec.begin(); 
  volume_stream.setVolume(0.10);    // Set volume
  while (dec.copy());                     // Decode and send to DAC
}
void loop() {
  File music_file = music_dir.openNextFile();
  if (music_file) { 
    Serial.print("Now playing: "); 
    Serial.println(music_file.name()); 
    play(music_file); 
  } else {
    while(1);   // No more music to play
  }
}
I get this output: 
09:41:41.376 -> [I] VolumeStream.h : 199 - setVolume: 1.000000 at 1
09:41:41.376 -> [I] AudioTypes.h : 127 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:41:41.376 -> [I] AudioTypes.h : 127 -  sample_rate: 44100 / channels: 2 / bits_per_sample: 16
09:41:41.376 -> [I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
09:41:41.376 -> [I] I2SConfigESP32.h : 81 - port_no: 0
09:41:41.376 -> [I] I2SConfigESP32.h : 82 - is_master: Master
09:41:41.376 -> [I] I2SConfigESP32.h : 83 - sample rate: 44100
09:41:41.376 -> [I] I2SConfigESP32.h : 84 - bits per sample: 16
09:41:41.376 -> [I] I2SConfigESP32.h : 85 - number of channels: 2
09:41:41.376 -> [I] I2SConfigESP32.h : 86 - signal_type: Digital
09:41:41.376 -> [I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
09:41:41.376 -> [I] I2SConfigESP32.h : 90 - auto_clear: true
09:41:41.376 -> [I] I2SConfigESP32.h : 92 - use_apll: true
09:41:41.376 -> [I] I2SConfigESP32.h : 97 - buffer_count:6
09:41:41.376 -> [I] I2SConfigESP32.h : 98 - buffer_size:512
09:41:41.376 -> [I] I2SConfigESP32.h : 103 - pin_bck: 14
09:41:41.376 -> [I] I2SConfigESP32.h : 105 - pin_ws: 15
09:41:41.376 -> [I] I2SConfigESP32.h : 107 - pin_data: 22
09:41:41.410 -> Now playing: /Music/03 Toto - Georgy Porgy.flac
09:41:41.410 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:41:41.410 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:41:41.410 -> [I] CodecFLAC.h : 85 - FLAC is open
09:41:41.410 -> [I] VolumeStream.h : 199 - setVolume: 0.100000 at 0
09:41:41.410 -> [I] VolumeStream.h : 199 - setVolume: 0.100000 at 1
09:45:59.879 -> [E] CodecFLAC.h : 109 - not active
09:45:59.946 -> Now playing: /Music/09 Toto - Hold The Line.flac
09:45:59.946 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:45:59.946 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:45:59.946 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:45:59.946 -> [E] CodecFLAC.h : 109 - not active
09:45:59.946 -> Now playing: /Music/01 Talking Heads - Psycho Killer.flac
09:45:59.946 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:45:59.946 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:45:59.946 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:45:59.946 -> [E] CodecFLAC.h : 109 - not active
09:45:59.946 -> Now playing: /Music/06 Steely Dan - Hey Nineteen.flac
09:45:59.946 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:45:59.946 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:45:59.946 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:45:59.946 -> [E] CodecFLAC.h : 109 - not active
09:45:59.946 -> Now playing: /Music/01 Dire Straits - So Far Away.flac
09:45:59.946 -> [I] CodecFLAC.h : 60 - virtual bool audio_tools::FLACDecoder::begin()
09:45:59.980 -> [I] CodecFLAC.h : 70 - FLAC__stream_decoder_new
09:45:59.980 -> [E] CodecFLAC.h : 81 - ERROR: initializing decoder: FLAC__STREAM_DECODER_INIT_STATUS_ALREADY_INITIALIZED
09:45:59.980 -> [E] CodecFLAC.h : 109 - not active

This is different to your output.  
I am using the latest zip from github and the begin() function in CodecFLAC.h is: 
  bool begin() {
    TRACEI();
    is_active = false;
    if (decoder == nullptr) {
      if ((decoder = FLAC__stream_decoder_new()) == NULL) {
        LOGE("ERROR: allocating decoder");
        is_active = false;
        return false;
      }
      LOGI("FLAC__stream_decoder_new");
    }

    // if it is already active we close it
    auto state = FLAC__stream_decoder_get_state(decoder);
    if (state != FLAC__STREAM_DECODER_UNINITIALIZED){
      FLAC__stream_decoder_finish(decoder);
    }

    // deactivate md5 checking
    FLAC__stream_decoder_set_md5_checking(decoder, is_md5_checing);

    // init decoder
    if (is_ogg){
      init_status = FLAC__stream_decoder_init_ogg_stream( decoder, read_callback, nullptr, nullptr, nullptr, nullptr, write_callback, nullptr, error_callback, this);
    } else {
      init_status = FLAC__stream_decoder_init_stream( decoder, read_callback, nullptr, nullptr, nullptr, nullptr, write_callback, nullptr, error_callback, this);
    }

    if (init_status != FLAC__STREAM_DECODER_INIT_STATUS_OK) {
      LOGE("ERROR: initializing decoder: %s", FLAC__StreamDecoderInitStatusString[init_status]);
      is_active = false;
      return false;
    }
    LOGI("FLAC is open");
    is_active = true;
    return true;
  }

Any idea what is different?  
NB: I am using an old arduino core.  Board is ESP32 Huzzah32.

**Comment by pschatzmann** [comment]
The line numbers of the logging does not match with the actual code.
So you must be using some old library version somewhere...

**Reply by geoffw64** [reply]
Correct.  Deleted the libraries, downloaded and re-compiled - now works.  
Thanks for your help Phil.  :)
